plugin/pkg/scheduler/algorithm/priorities/priorities.go

```diff
@@ -277,3 +277,79 @@ func fractionOfCapacity(requested, capacity int64) float64 {
 	}
 	return float64(requested) / float64(capacity)
 }
+
+type NodePreferAvoidPod struct {
+	controllerLister algorithm.ControllerLister
+	replicaSetLister algorithm.ReplicaSetLister
+}
+
+func NewNodePreferAvoidPodsPriority(controllerLister algorithm.ControllerLister, replicaSetLister algorithm.ReplicaSetLister) algorithm.PriorityFunction {
+	nodePreferAvoid := &NodePreferAvoidPod{
+		controllerLister: controllerLister,
+		replicaSetLister: replicaSetLister,
+	}
+	return nodePreferAvoid.CalculateNodePreferAvoidPodsPriority
+}
+
+func (npa *NodePreferAvoidPod) CalculateNodePreferAvoidPodsPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
```

Comment:
So let's do this way:
1) let's merge as it is now (basically in selector spreading we are still calling all the functions that we have here, so it shouldn't be worse)
2) Once this is merged, I will do a pass over it and fix some issues that I already fixed for other priorities/predicates
3) once ControllerRef is ready, I will switch to that (together with selector_spreading)

[Generally, the asymptotical complexity of it is the same as selector spreading, We should probably solve both of them in the same way in the future.]